### ISSUE TITLE
Should regex-based .matches() be more symmetric with .replace()?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#21931]
I am admittedly not a heavy or regular user of regex, and the following question/issue should take that into account.

While updating codes to reflect changes to regex in Chapel 1.30, I generally liked the changes from `.sub()` and `compile()` to `.replace()` and `new regex()`.  For example, here is one such change, which I think is much more readable in its current form:

```diff
-        copy = compile(f).sub(r, copy);
+        copy = copy.replace(new regex(f), r);
```

In another place, my code used `matches()` which looked like:

```chapel
for m in compile(pattern).matches(data) do ...
```

and now looks like:

```chapel
for m in (new regex(pattern)).matches(data) do ...
```

where I could obviously put in temps to capture the regex as follows:

```chapel
var myre = new regex(pattern);
for m in myre.matches(data) do ...
```

This made me wonder whether such cases should be written as:

```chapel
var myre = new regex(pattern);
for m in data.matches(myre) do ...
```

for better symmetry with `.replace()`.  This would also support writing it more cleanly as a one-statement form, if desired:

```chapel
for m in data.matches(new regex(pattern)) do ...
```

In this issue I'm asking whether:
* this resonates with others?
* the existing asymmetry has a benefit that I'm not seeing?
* this proposal has (or doesn't have) precedent in other languages/regex systems?
* this is a change we'd want to make for Chapel 2.0?

**Comment by e-kayrakli** [comment]
First a clarification: are you proposing we deprecate `regex.matches(string)` in favor of `string.matches(regex)`?

I said these to Brad while we were chatting about this earlier, but I'll capture my thoughts here:

> this resonates with others?

This proposal does resonate with me but I have some reservations. First, I like the conciseness of the last snippet above. Second, I think it is an interesting approach to look at `regex` and `string`/`bytes` interfaces together while asking the question whether we can have overloads or new methods+iterators where such `regex`es are more of a first-class citizen. My reservations for this specific proposal are:

- fundamentally, this is proposing we reverse the receiver and argument of a method for which I think we have the right roles because "matching" is a pretty regex-specific operation to me. Thus having a "match" method or "matches" iterator on `string`/`bytes` feels a bit off. 
- `string.matches` as above is way too close to `string.find` in terms of what they do, I would prefer not to have two things with very different names doing very similar thing.

Based on those grounds, my counter proposal would be to have a new `string.findAll(regex(string))` iterator that has the proposed `matches` behavior as above. Note that

1. we already have `string.find(regex(string))` which returns a `byteIndex`
2. we can still keep `regex.matches(string)` as a `regexMatch` yielding alternative as opposed to `findAll` yielding `byteIndex`, which makes this counter-proposal strictly additive.
3. we can probably have two overloads of `findAll` with `string` and `regex(string)` arguments.

But, I am not sure if the usefulness lost by returning `byteIndex` instead of a `regexMatch` is a deal breaker.

> this is a change we'd want to make for Chapel 2.0?

You can probably guess, but my stance is "why rush?".

**Comment by bradcray** [comment]
> First a clarification: are you proposing we deprecate regex.matches(string) in favor of string.matches(regex)?

I'm posing the question to see whether my viewpoint as a regex-naive developer resonates with those who are more expert here or not.  If the reaction is "no—there are good reasons we have what we have today / wouldn't want to do this" (e.g., "it's what other popular regex libraries do") then I'd propose we close this issue instead.

> my counter proposal would be to have a new string.findAll(regex(string)) iterator that has the proposed matches behavior as above.

That'd be fine with me as well.  I'm not particularly attached to a specific name and agree that `findAll()` is more symmetric name-wise with `find()`.  Does the fact that `find()` returns a `byteIndex` while `findAll()`/`matches()` yields `regexMatch`es be weird?  (and/or, is it at all weird that `myString.find(myRegex)` doesn't return a `regexMatch()`?)

>  this is proposing we reverse the receiver and argument of a method for which I think we have the right roles because "matching" is a pretty regex-specific operation to me. Thus having a "match" method or "matches" iterator on string/bytes feels a bit off.

My non-expert observation here is that it feels like a lot of the regex-oriented routines I've come into contact with recently (`.readThrough/To()`, `.advanceThrough/To()`, `.replace()`) take regexes as arguments rather than having them be methods on the regexes themselves.  That said, there may be many other similar features that are methods on regexes and I just haven't encountered them yet or noticed them recently.

> You can probably guess, but my stance is "why rush?".

So we can stabilize Regex for 2.0 and not keep changing user codes that make use of the features?

**Comment by e-kayrakli** [comment]
> So we can stabilize Regex for 2.0 and not keep changing user codes that make use of the features?

Yes. The question in this topic doesn't make me worried about what we have today enough to make this a 2.0 discussion, personally. IOW, what we have today in `re.matches(string)` feels natural to me.


---

I am also wondering, whether regex literals (https://github.com/chapel-lang/chapel/issues/17219) can make the little roundabout-ness in the OP. i.e., if `pattern` were to be declared as something like `var pattern = r"foo";` it would have been equivalent to `var pattern = new regex("foo");` today. `for i in r"foo".matches(myString)` could be a thing, for example. I realize that it is a bit tangent, though.

**Comment by bradcray** [comment]
> The question in this topic doesn't make me worried about what we have today enough to make this a 2.0 discussion, personally.

I think you're saying "I think our decision for 2.0 should be that the current interface is good enough s.t. we'll commit to not changing / retiring it" rather than "This routine should be allowed to be deprecated/changed after 2.0", is that right?

> for i in r"foo".matches(myString) could be a thing, for example. I realize that it is a bit tangent, though.

To me, that still reads more awkwardly than

```chapel
for m in myString.matches(r"foo") ...
for m in myString.findAll(r"foo") ...
```

But I agree that this is tangential, in that it's the inversion between `.matches()` and `.replace()` over which is the receiver and which is the argument that gives me pause, not the syntax used to create the regular expression (though I do think `new regex()` was a strict improvement over `compile()` and agree that `r"..."` could be even more attractive if we went that route).

**Comment by e-kayrakli** [comment]
> I think you're saying "I think our decision for 2.0 should be that the current interface is good enough s.t. we'll commit to not changing / retiring it" rather than "This routine should be allowed to be deprecated/changed after 2.0", is that right?

Yes.

**Comment by lydia-duncan** [comment]
It looks like the decision here was "if we change anything, it won't be in a way that breaks what is already stabilized".  I think that means I can remove the 2.0 tag, does that sound right @bradcray and @e-kayrakli?

**Comment by e-kayrakli** [comment]
That sounds right to me.
